doc: improve example for Error.captureStackTrace()

brodo opened the pull request:
Change the `MyError` example so that instances of `MyError` are `instanceof Error` and also  native errors when checked with `util.types.isNativeError()`.

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

BridgeAR commented on doc/api/errors.md:
````diff
@@ -233,7 +233,9 @@ details of error generation from the user. For instance:
 ```js
 function MyError() {
-  Error.captureStackTrace(this, MyError);
+  const instance = new Error()
````
```suggestion
  const { stackTraceLimit } = Error
  Error.stackTraceLimit = 0
  const instance = new Error()
  Error.stackTraceLimit = stackTraceLimit
```

Without this, the stack frames would be computed twice. That would be a lot of overhead that is not needed. Creating errors is actually quite expensive CPU wise.

brodo replied:
Good to know! I'll add a comment that explains this because otherwise I think it's a little confusing.

BridgeAR reviewed (commented):
I see why you suggest this. I am not certain if the example becomes more difficult to read though, especially as it has new implications as outlined in my comment.

brodo commented:
Should we maybe add a suggestion to use `class MyError extends Error {}`? Instances of this class will be `instanceof Error`, native Error and will not show the MyError constructor in the stack trace.

brodo commented:
> I see why you suggest this. I am not certain if the example becomes more difficult to read though, especially as it has new implications as outlined in my comment.

I understand the readability concern. Essentially it's a trade-off between having a 'copyable' example and a readable one. It would be interesting to find out if there are packages on rpm that contain this error implementation...

BridgeAR commented on doc/api/errors.md:
````diff
@@ -233,7 +233,15 @@ details of error generation from the user. For instance:
 ```js
 function MyError() {
-  Error.captureStackTrace(this, MyError);
+  // To avoid calculating the stack trace twice, the stack trace limit is
+  // to zero. Before that, the current stack trace limit is saved so it can be
+  // restored later.
````
```suggestion
  // Set the stack trace limit to zero to avoid calculating the stack trace twice.
```

BridgeAR commented on doc/api/errors.md:
````diff
@@ -242,6 +250,15 @@ function MyError() {
 new MyError().stack;
 ```
 
+We recommend using classes that extend `Error` for custom errors:
````
Error.captureStackTrace may even be used in combination with custom errors. It's usage is mainly to avoid specific stack frames. It's therefore not really an alternative to classes.

brodo replied:
Oh right, that makes sense. I've only thought about hiding constructors, which is not necessary when you subclass `Error`. I'll just get rid of the whole 'We recommend...' part including the code example.

BridgeAR reviewed (commented):
I am hesitant to change the documentation like that as we mix different APIs and errors should ideally not be created like that in the first place.

What about changing the example to show how to hide multiple call sites? E.g.,

```js
function a() {
  b();
}

function b() {
  c();
}

function c() {
  // Create an error without stack trace to avoid calculating the stack trace twice.
  const { stackTraceLimit } = Error;
  Error.stackTraceLimit = 0;
  const error = new Error();
  Error.stackTraceLimit = stackTraceLimit;

  // Capture the stack trace above function b
  Error.captureStackTrace(error, b); // Neither function c, nor b is included in the stack trace
  throw new Error();
}
```

That way we create a "real" error and only change it's stack trace accordingly.

brodo commented:
> ```js
> function a() {
>   b();
> }
> 
> function b() {
>   c();
> }
> 
> function c() {
>   // Create an error without stack trace to avoid calculating the stack trace twice.
>   const { stackTraceLimit } = Error;
>   Error.stackTraceLimit = 0;
>   const error = new Error();
>   Error.stackTraceLimit = stackTraceLimit;
> 
>   // Capture the stack trace above function b
>   Error.captureStackTrace(error, b); // Neither function c, nor b is included in the stack trace
>   throw new Error();
> }
> ```

I like this example a lot. It it illustrated the usage of the  function much better.

debadree25 commented:
Could you update the first commit message, it seems the description exceeds the size limit of 72

debadree25 reviewed (approved):
The example looks good, would also wait for @BridgeAR 's review to see if the changes address all his concerns.

brodo commented:
> Could you rebase and update the first commit message, it seems the description exceeds the size limit of 72

I've squashed all the commits and added line breaks.
